File: 프로그래머스/2/42584. 주식가격/주식가격.py
def solution(prices):
    length = len(prices)
    answer = [0] * length 
    
    stack = [] 

    for i, price in enumerate(prices):
        while stack and prices[stack[-1]] > price:
            past_time = stack.pop() 
            answer[past_time] = i - past_time 
            
        stack.append(i)

    while stack:
        past_time = stack.pop()
        answer[past_time] = length - 1 - past_time

    return answer

File: 프로그래머스/2/42584. 주식가격/주식가격.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;

vector<int> solution(vector<int> prices) {
    int n = prices.size();
    vector<int> answer(n);
    stack<int> s;   

    for (int i = 0; i < n; i++) {
        while (!s.empty() && prices[s.top()] > prices[i]) {
            int past_time = s.top();
            s.pop();
            answer[past_time] = i - past_time; 
        }
        s.push(i);
    }

    while (!s.empty()) {
        int past_time = s.top();
        s.pop();
        answer[past_time] = n - 1 - past_time;
    }

    return answer;
}